core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala

```diff
@@ -107,10 +109,20 @@ class InvokerReactive(
     }
 
     /** Sends an active-ack. */
-    val ack = (tid: TransactionId, activation: WhiskActivation) => {
+    val ack = (tid: TransactionId, activationResult: WhiskActivation) => {
         implicit val transid = tid
-        producer.send("completed", CompletionMessage(tid, activation, s"invoker$instance")).andThen {
-            case Success(_) => logging.info(this, s"posted completion of activation ${activation.activationId}")
+
+        def send(res: Either[ActivationId, WhiskActivation], recovery: Boolean = false) = {
+            val msg = CompletionMessage(transid, res, this.name)
+            producer.send("completed", msg).andThen {
+                case Success(_) =>
+                    logging.info(this, s"posted ${if (recovery) "recovery" else ""} completion of activation ${activationResult.activationId}")
+            }
+        }
+
+        send(Right(activationResult)).onFailure {
+            case t if t.getCause.isInstanceOf[RecordTooLargeException] =>
+                send(Left(activationResult.activationId), recovery = true)
```

Comment:
You'll see this later. I think it is ok in that it's a recovery for a specific failure case not all Kafka errors.